src/Microsoft.IdentityModel.JsonWebTokens/JsonWebToken.cs

```diff
@@ -67,434 +113,658 @@ public JsonWebToken(string header, string payload)
+            if (Dot3 == -1)
             {
-                Header = JObject.Parse(Base64UrlEncoder.Decode(tokenParts[0]));
+                IsSigned = !(Dot2 + 1 == encodedJson.Length);
+                try
+                {
+#if NET45
+                    _sChars = IsSigned ? encodedJson.ToCharArray(Dot2 + 1, encodedJson.Length - Dot2 - 1) : string.Empty.ToCharArray();
+                    SignatureBytes = Base64UrlEncoder.UnsafeDecode(_sChars);
+                    _hChars = encodedJson.ToCharArray(0, Dot1);
+                    Header = new JsonClaimSet(Base64UrlEncoder.UnsafeDecode(_hChars));
+#else
+                    Header = new JsonClaimSet(JwtTokenUtilities.GetJsonDocumentFromBase64UrlEncodedString(encodedJson, 0, Dot1));
+#endif
+                }
+                catch (Exception ex)
+                {
+                    throw LogHelper.LogExceptionMessage(new ArgumentException(LogHelper.FormatInvariant(LogMessages.IDX14102, encodedJson.Substring(0, Dot1), encodedJson), ex));
+                }
+
+                try
+                {
+#if NET45
+                    MessageBytes = Encoding.UTF8.GetBytes(encodedJson.ToCharArray(0, Dot2));
+                    _pChars = encodedJson.ToCharArray(Dot1 + 1, Dot2 - Dot1 - 1);
+                    Payload = new JsonClaimSet(Base64UrlEncoder.UnsafeDecode(_pChars));
+#else
+                    Payload = new JsonClaimSet(JwtTokenUtilities.GetJsonDocumentFromBase64UrlEncodedString(encodedJson, Dot1 + 1, Dot2 - Dot1 - 1));
+#endif
+                }
+                catch (Exception ex)
+                {
+                    throw LogHelper.LogExceptionMessage(new ArgumentException(LogHelper.FormatInvariant(LogMessages.IDX14101, encodedJson.Substring(Dot2, Dot2 - Dot1), encodedJson), ex));
+                }
             }
-            catch (Exception ex)
+            else
             {
-                throw LogHelper.LogExceptionMessage(new ArgumentException(LogHelper.FormatInvariant(LogMessages.IDX14102, tokenParts[0], rawData), ex));
+                // empty payload for JWE's {encrypted tokens}.
+#if NET45
+                Payload = new JsonClaimSet("{}");
+#else
+                Payload = new JsonClaimSet(JsonDocument.Parse("{}"));
+#endif
+                if (Dot3 == encodedJson.Length)
+                    throw LogHelper.LogExceptionMessage(new ArgumentException(LogHelper.FormatInvariant(LogMessages.IDX14100, encodedJson)));
+
+                Dot4 = encodedJson.IndexOf('.', Dot3 + 1);
+
+                // JWE needs to have 4 dots
+                if (Dot4 == -1 || Dot4 == encodedJson.Length)
+                    throw LogHelper.LogExceptionMessage(new ArgumentException(LogHelper.FormatInvariant(LogMessages.IDX14100, encodedJson)));
+
+                // too many dots...
+                if (encodedJson.IndexOf('.', Dot4 + 1) != -1)
+                    throw LogHelper.LogExceptionMessage(new ArgumentException(LogHelper.FormatInvariant(LogMessages.IDX14100, encodedJson)));
```

Comment:
same comment here, would be nice to differentiate from PII scrubbed logs which error case we hit for quicker troubleshooting. At least differentiating between encrypted and non-encrypted would be helpful.